dbms/src/Storages/DeltaMerge/SSTFilesToDTFilesOutputStream.h

```diff
@@ -56,52 +51,131 @@ enum class FileConvertJobType
+     */
+    SSTFilesToDTFilesOutputStream(ChildStream child_,
                                   StorageDeltaMergePtr storage_,
                                   DecodingStorageSchemaSnapshotConstPtr schema_snap_,
                                   TiDB::SnapshotApplyMethod method_,
                                   FileConvertJobType job_type_,
-                                  TMTContext & tmt_);
+                                  UInt64 split_after_rows_,
+                                  UInt64 split_after_size_,
+                                  Context & context);
     ~SSTFilesToDTFilesOutputStream();
 
     void writePrefix();
     void writeSuffix();
     void write();
 
+    /**
+     * The DTFile page ids that can be ingested.
+     * The returned vector is ensured to be ordered in ascending order.
+     */
     PageIds ingestIds() const;
 
     // Try to cleanup the files in `ingest_files` quickly.
     void cancel();
 
 private:
+    /**
+     * Generate a DMFilePtr and its DMFileBlockOutputStream.
+     */
     bool newDTFileStream();
-
-    // Stop the process for decoding committed data into DTFiles
-    void stop();
+    /**
+     * Close the current DMFile stream.
+     */
+    bool finalizeDTFileStream();
 
 private:
-    BoundedSSTFilesToBlockInputStreamPtr child;
+    ChildStream child;
     StorageDeltaMergePtr storage;
     DecodingStorageSchemaSnapshotConstPtr schema_snap;
     const TiDB::SnapshotApplyMethod method;
     const FileConvertJobType job_type;
-    TMTContext & tmt;
-    Poco::Logger * log;
+    const UInt64 split_after_rows;
+    const UInt64 split_after_size;
+    Context & context;
+    LoggerPtr log;
 
     std::unique_ptr<DMFileBlockOutputStream> dt_stream;
 
     std::vector<DMFilePtr> ingest_files;
 
-    size_t schema_sync_trigger_count = 0;
-    size_t commit_rows = 0;
+    /**
+     * How many rows has been committed to the current DTFile.
+     */
+    size_t committed_rows_this_dt_file = 0;
+    size_t committed_bytes_this_dt_file = 0;
+
+    /**
+     * How many rows has been committed so far.
+     */
+    size_t total_committed_rows = 0;
+    size_t total_committed_bytes = 0;
+
     Stopwatch watch;
 };
 
+class MockSSTFilesToDTFilesOutputStreamChild : private boost::noncopyable
+{
+public:
+    MockSSTFilesToDTFilesOutputStreamChild(BlockInputStreamPtr mock_data_, RegionPtr mock_region_) //
+        : mock_data(mock_data_)
+        , mock_region(mock_region_)
+    {}
+
+    void readPrefix()
+    {
+        mock_data->readPrefix();
+    }
+
+    void readSuffix()
+    {
+        mock_data->readSuffix();
+    }
+
+    RegionPtr getRegion() const
+    {
+        return mock_region;
+    }
+
+    Block read()
+    {
+        return mock_data->read();
+    }
+
+    std::tuple<size_t, size_t, size_t, UInt64> getMvccStatistics() const
+    {
+        return {};
+    }
+
+    SSTFilesToBlockInputStream::ProcessKeys getProcessKeys() const
+    {
+        return {};
+    }
+
+protected:
```

Comment:
Why we need protected here?

Reply:
It should be sufficient for current test cases to not accessing them. So it is made protected / private.

Reply:
I think private is ok. But not a big problem